### PR TITLE
Refactor downloads, add download interface

### PR DESCRIPTION
**Pull request by bcai2**
Adds:
* `toolchest.download(output_path, s3_uri="s3://your-s3-uri")` and `output.download(output_dir)`, user-invoked functions which can download an output after its job has completed.
* `ToolchestDownloadException`, a new exception thrown whenever an error is encountered in downloading or unpacking.
* Integration tests for manual downloads.

Modifies:
* Code for downloading has now been moved from `Query` to `api/downloads.py`, which serves as a single source for all download functionality.
* Downloads during `Query.run_query()` now call the functions in `api/downloads.py`.
* URLs used for the API have been moved to `api/urls.py`, which serves as a single source for all API URLs.
* All `toolchest.tool()` calls now correctly return an `Output`.

Removes:
* `presigned_s3_url` parameter for `Output`

### CODE REVIEW

**Comment by bcai2** [comment]
@lebovic Ready to review, whenever you're at a good stopping point. Otherwise, I'll pre-emptively merge this into staging and respond to changes in a new PR.

File: tests/test_download.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+toolchest_api_key = os.environ.get("TOOLCHEST_API_KEY")
+if toolchest_api_key:
+    toolchest.set_key(toolchest_api_key)
+
+KRAKEN2_SINGLE_END_HASH = 886254946
+
+
+@pytest.mark.integration
+def test_kraken2_output_manual_download():
```
Comment (lebovic) [inline tests/test_download.py:15]: About how long does this test take?

File: toolchest_client/api/download.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+from toolchest_client.api.auth import get_headers
+from toolchest_client.api.exceptions import ToolchestDownloadError
+from toolchest_client.api.urls import PIPELINE_URL
+from toolchest_client.files import get_file_type, get_params_from_s3_uri, unpack_files
+
+
+def download(output_path, s3_uri=None, pipeline_segment_instance_id=None,
+             output_file_keys=None, output_type=None, ):
+    """Downloads output to ``output_path``.
```
Comment (lebovic) [inline toolchest_client/api/download.py:27]: Possible nit: unintentional "``"?

File: toolchest_client/api/download.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+            # Note: this assumes the pipeline segment instance ID is embedded in the egress S3 URI.
+            # This supersedes pipeline_segment_instance_id, if it is provided.
+            s3_uri_params = get_params_from_s3_uri(s3_uri)
+            pipeline_segment_instance_id = s3_uri_params["key_initial"]
+        if pipeline_segment_instance_id:
+            output_s3_uri, output_file_keys = get_download_details(pipeline_segment_instance_id)
+        else:
+            error_message = "S3 URI of output not provided."
+            raise ToolchestDownloadError(error_message) from None
+    print(output_file_keys)
```
Comment (lebovic) [inline toolchest_client/api/download.py:58]: Nit: leftover `print`?

File: toolchest_client/files/tests/test_s3.py [hunk 4]
```diff
@@ -5,12 +5,14 @@
 
 
 def test_s3_params():
-    example_s3_uri = "s3://toolchest-public-examples/example.fastq"
+    example_s3_uri = "s3://toolchest-public-examples/dummy-id/example.fastq"
     params = get_params_from_s3_uri(example_s3_uri)
     target_params = {
-        "arn": "arn:aws:s3:::toolchest-public-examples/example.fastq",
+        "arn": "arn:aws:s3:::toolchest-public-examples/dummy-id/example.fastq",
```
Comment (lebovic) [inline toolchest_client/files/tests/test_s3.py:11]: Worth noting that we're now expecting the uuid to be present in the S3 URI structure. This isn't reflected in the db schema rn, where the uuid is just a part of the file prefix.

**Review by lebovic — APPROVED** [review]
👍

File: toolchest_client/api/download.py (hunk 2 above)
Comment (bcai2) [reply toolchest_client/api/download.py:27]: That was a holdover from the previous comment in `query.py`, but it does seem unnecessary, so I'll get rid of it.

File: toolchest_client/api/download.py (hunk 3 above)
Comment (bcai2) [reply toolchest_client/api/download.py:58]: Yup, good catch

File: tests/test_download.py (hunk 1 above)
Comment (bcai2) [reply tests/test_download.py:15]: Not too much longer than the `kraken2` single-end run (~3 mins). This is basically just the same as the normal `kraken2` single-end run, but downloading the output in 3 different ways after the job finishes.

**Pull request merged**